### PR TITLE
1.0 dev

### PR DESCRIPTION
**Pull request by scinos**
Major refactor, preparing for v1.0:

### Breaking changes

- Simplified CLI. Renamed the main tool from `yarn-tools` to `yarn-dedupe`. Now everything is done with one command:
```
node cli.js --help

  Usage: cli [options] <yarn.lock path>


  Options:

    -V, --version              output the version number
    -s, --strategy <strategy>  Deduplication strategy. Valid values: fewer, highest. Default is "highest"
    -l, --list                 Do not change yarn.lock, just output the diagnosis
    --packages <packages>      A comma separated list of packages to dedupe. Defauls to all packages.
    --print                    Instead of saving the deduped yarn.lock, print the result in stdout
    -h, --help                 output usage information
```

- Added an option to specify the strategy when deduping files: `highest` and `fewer` are valid strategies.

### Other changes
- Better support for versions like `<package>` (without the `@<version>` part).
- Better support for "invalid" versions like `sinon@next`.
- Added an option to save `yarn.lock` instead of printing the result to the console.
- Separated lib logic into `index.js`, and created `cli.js` for handling the command line options.

### CODE REVIEW

**Comment by scinos** [comment]
@bj00rn @gfx @mxmul @lukebatchelor 

Hey folks. I did a huge refactor of this library, preparing for 1.0. Would you be able to take a look at the PR? Or check that this new version still works for your projects?

**Comment by lukebatchelor** [comment]
Will do in the morning. 

Should we call out the breaking change of writing in place by default too?

**Comment by scinos** [comment]
@lukebatchelor 
Added more info to the Changelog and a Migration Guide to the README

**Comment by lukebatchelor** [comment]
Sweet, tested in the atlaskit repo using yarn 1.12.0.

All commands and flags working as expected. However, when I deduped, committed and ran yarn again, it caused some changes. I believe I saw you mention this at some point, but I don't recall where...

It's not a huge issue, the things that it changed were not undoing any of the deduplication (running dedupe a second time after yarn changes doesn't report any changes). It looks like it's just removed a bunch of unused packages (I believe this is the same behaviour you saw too?).

There were a couple of tiny typos (I'll put comments), but nothing to block a release imo.

File: cli.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+const {fixDuplicates, listDuplicates} = require('./index');
+const version = require('./package.json').version;
+
+commander
+    .version(version)
+    .usage('[options] <yarn.lock path>')
+    .option('-s, --strategy <strategy>', 'Deduplication strategy. Valid values: fewer, highest. Default is "highest"', /^(fewer|highest)$/i, 'highest')
+    .option('-l, --list', 'Do not change yarn.lock, just output the diagnosis')
+    .option('--packages <packages>', 'A comma separated list of packages to dedupe. Defauls to all packages.', val => val.split(',').map(v => v.trim()))
+    .option('--print', 'Instead of saving the deduped yarn.lock, print the result in stdout')
```
Comment (lukebatchelor) [inline cli.js]: Super nipick, but the two extra options that commander adds, `--version` and `--help` start their sentences with lower case and look out of place. Might be worth making them all lowercase?

image

File: cli.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+const {fixDuplicates, listDuplicates} = require('./index');
+const version = require('./package.json').version;
+
+commander
+    .version(version)
+    .usage('[options] <yarn.lock path>')
+    .option('-s, --strategy <strategy>', 'Deduplication strategy. Valid values: fewer, highest. Default is "highest"', /^(fewer|highest)$/i, 'highest')
+    .option('-l, --list', 'Do not change yarn.lock, just output the diagnosis')
+    .option('--packages <packages>', 'A comma separated list of packages to dedupe. Defauls to all packages.', val => val.split(',').map(v => v.trim()))
```
Comment (lukebatchelor) [inline cli.js]: Typo
```suggestion
    .option('--packages <packages>', 'A comma separated list of packages to dedupe. Defaults to all packages.', val => val.split(',').map(v => v.trim()))
```

**Comment by bj00rn** [comment]
> @bj00rn @gfx @mxmul @lukebatchelor
> 
> Hey folks. I did a huge refactor of this library, preparing for 1.0. Would you be able to take a look at the PR? Or check that this new version still works for your projects?

Seems to be working fine, except for some broken package.json files in .history which breaks flow type checking in my project.  https://github.com/atlassian/yarn-deduplicate/issues/14

**Pull request merged**